From ca0c4ee7cc61bb989238054ca6171b7ea049de7c Mon Sep 17 00:00:00 2001
From: Suneel Garapati <sgarapati@marvell.com>
Date: Thu, 21 Nov 2019 15:32:01 -0800
Subject: [PATCH 0438/1921] spi: octeontx2: use read after write for MPI_CFG

Use read to make previous write observed for mode change
in config register.
Also, make delay after mode change in MPI_CFG register
to be module parameter.

Change-Id: I96d52e6506a2dffc0c6895675ca524d0d24ee319
Signed-off-by: Suneel Garapati <sgarapati@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/19111
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
(cherry picked from commit 6e6ab555da18884ccf153cddaec1a925d26a8180)
Reviewed-on: https://sj1git1.cavium.com/19140
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26933
Reviewed-by: Suneel Garapati <sgarapati@caviumnetworks.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/spi/spi-octeontx2.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/spi/spi-octeontx2.c b/drivers/spi/spi-octeontx2.c
index e797b00e9e61..c98b5c287b8b 100644
--- a/drivers/spi/spi-octeontx2.c
+++ b/drivers/spi/spi-octeontx2.c
@@ -23,6 +23,11 @@ module_param(tbi_clk_en, uint, 0644);
 MODULE_PARM_DESC(tbi_clk_en,
 		 "Use Fixed Time Base 100MHz Reference Clock (0=Disable, 1=Enable [default])");
 
+static int cfg_mode_delay = 30;
+module_param(cfg_mode_delay, uint, 0644);
+MODULE_PARM_DESC(cfg_mode_delay,
+		 "Delay in micro-seconds for mode change in MPI CFG register (30 [default])");
+
 static void octeontx2_spi_wait_ready(struct octeontx2_spi *p)
 {
 	union mpix_sts mpi_sts;
@@ -105,7 +110,8 @@ static int octeontx2_spi_do_transfer(struct octeontx2_spi *p,
 	if (mpi_cfg.u64 != p->last_cfg) {
 		p->last_cfg = mpi_cfg.u64;
 		writeq(mpi_cfg.u64, p->register_base + OCTEONTX2_SPI_CFG(p));
-		udelay(100); /* allow CS change to settle */
+		mpi_cfg.u64 = readq(p->register_base + OCTEONTX2_SPI_CFG(p));
+		udelay(cfg_mode_delay); /* allow CS change to settle */
 	}
 	tx_buf = xfer->tx_buf;
 	rx_buf = xfer->rx_buf;
-- 
2.31.1

